### ISSUE TITLE
Floki.find returning unmatching nodes

### ISSUE THREAD

**Issue by mmmries** [philss/floki#84]
I tried to update a project to the new `0.12.0` release, but ran into a bug. Here is a minimal reproduction of the error:

```elixir
    html = """
      <html>
      <body>
      <div id="messageBox" class="legacyErrors"><div class="messageBox error"><h2 class="accessAid">Error Message</h2><p>There has been an error in your account.</p></div></div>
      <div id="main" class="legacyErrors"><p>Separate Error Message</p></div>
      </body>
      </html>
      """
    html |> Floki.parse |> Floki.find(".messageBox p") |> Floki.text
    # => "There has been an error in your account.Separate Error Message"
```

The `.messageBox p` selector should only return the first `p` node, but it actually returns both of the `p` nodes and the `Floki.text` call ends up returning the text from both of those nodes.

**Comment by mmmries** [comment]
BTW thanks for doing this project. It has been very useful so far and I'd be happy to spend some time trying to fix this bug if you can point me in the right general direction of where I should look.

**Comment by philss** [comment]
Hi @mmmries! 

Thanks for the report. This is definitely a bug. Not sure why, but is selecting a wrong node.
I'm investigating this.

**Comment by philss** [comment]
@mmmries I found the problem, but I'm not sure how to fix yet. It is here: https://github.com/philss/floki/blob/master/lib/floki/finder.ex#L202-L210

We should take all the children ids based on the list of ids that are between the current node and the sibling node. This works when there is a sibling, but don't work when the node is the last node inside his parent.

I have one idea that I will try to implement. Feel free to give a try too.

Thanks!

**Comment by mmmries** [comment]
@philss thanks for pointing me in the right direction. I'll take a look and see if I can better understand that section of code specifically. Hopefully I can make an attempt at a fix before you get there so you can take me to 🏫

**Comment by philss** [comment]
@mmmries nice! One way to make sure it takes all descendant nodes is to traverse the nodes that are below the current node and adding them to the stack right in that process.
This would avoid the idea of taking the ids of what we assume it's a subtree, which theoretically is simpler. But would be a little bit harder to implement since we would have to traverse and accumulate the results.

**Comment by philss** [comment]
@mmmries This fix was released in v0.12.1. Thank you again!

**Comment by mmmries** [comment]
Thanks for the release and pointing me in the right direction to work on a fix :blue_heart: :yellow_heart: :green_heart:
